### PR TITLE
feat: add lazy credential manager

### PR DESCRIPTION
**Pull request by jquense**
Adds a new Lazy JWT manager that tries to update credentials only as needed. The idea is this should enable us to issue shorter lived credentials for long lived socket connections since re-up on the `exp` would happen lazily

### CODE REVIEW

File: src/LazyJwtCredentialsManager.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import { CredentialsManager } from './CredentialsManager';
+import { JwtCredentials } from './JwtCredentialsManager';
+
+const SECONDS_TO_MS = 1000;
+
+export default abstract class LazyJwtCredentialsManager<
```
Comment (itajaja) [inline src/LazyJwtCredentialsManager.ts:6]: does it make sense to share some code with the `JwtCredentialsManager`?

File: src/LazyJwtCredentialsManager.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  private isExpired(credentials: TCredentials) {
+    return Date.now() >= credentials.exp * SECONDS_TO_MS;
+  }
+
+  async getCredentials(): Promise<TCredentials | null | undefined> {
+    const credentials = await this.credentialsPromise;
+
+    if (credentials && this.isExpired(credentials)) {
```
Comment (itajaja) [inline src/LazyJwtCredentialsManager.ts:25]: uhm, so this acts as a lock, basically? trying to understand if there are instances where `updateCredentials` could be called twice

File: src/LazyJwtCredentialsManager.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+  abstract getCredentialsFromAuthorization(
+    authorization: string,
+  ):
+    | (TCredentials | null | undefined)
+    | Promise<TCredentials | null | undefined>;
+
+  async authenticate(token: string) {
+    this.token = token;
+    await this.updateCredentials();
```
Comment (itajaja) [inline src/LazyJwtCredentialsManager.ts:40]: do you need to call it here? I guess doesn't hurt?

**Review by itajaja — COMMENTED** [review]
this all makes sense, and I think that the lazy manager is a more sensical implementation, but I am not sure I understand why you said that the existing manager doesn't work. can you explain? Moreover, they seem to be functionally very similar, so do we need to have both?

File: src/LazyJwtCredentialsManager.ts (hunk 3 above)
Comment (jquense) [reply src/LazyJwtCredentialsManager.ts:40]: you need it to pull the credentials from the token now that it exists

**Comment by jquense** [comment]
We probably don't need both, this seems like the better implementation i think.

> you said that the existing manager doesn't work

Yeah that was an over statement, it works in the sense that it's valuable when your implementation of `getCredentialsFromAuthorization` produces something with a shorter `exp` than the auth token. It "doesn't work" in common case where you aren't doing anything fancy and simply decoding the JWT. In that case the code will "refresh" aster the token expires but not result in any new credentials, and if you `authenticate` with a new token you don't need the refresh logic. 

That's all to say if you are just using the jwt from the headers, you don't need refresh logic

File: src/LazyJwtCredentialsManager.ts (hunk 2 above)
Comment (jquense) [reply src/LazyJwtCredentialsManager.ts]: the stored promise is some manner of lock, it's not great one but I think it's fine if updateCredentials is called multiple times?

**Comment by jquense** [comment]
ok I replaced the old implementation instead of adding a new one

**Review by AaronBuxbaum — APPROVED** [review]

File: src/LazyJwtCredentialsManager.ts (hunk 2 above)
Comment (itajaja) [reply src/LazyJwtCredentialsManager.ts]: I think so!

**Review by itajaja — COMMENTED** [review]
great! seems not breaking, but confirm that

**Comment by jquense** [comment]
I guess it minorly breaking only in that the behavior changes and the constructor args are slightly different. It should be dirt simple to migrate to tho

**Pull request merged**